ext/standaloneusers/Civi/Api4/Action/User/Login.php

```diff
@@ -92,6 +113,46 @@ public function _run(Result $result) {
 
   }
 
+  protected function createRememberJWT(array $pending): string {
+    $days = \Civi::settings()->get('standalone_mfa_remember') ?? 0;
+    if (!$days || empty($pending['rememberMe'])) {
+      // Feature disabled by admin or request.
+      return '';
+    }
+
+    $hashedPass = $this->getPassHash($pending['userID']);
+    if (!$hashedPass) {
+      // This should not happen. But if it does, it's very fishy and we should
+      // not support bypassing MFA.
+      return '';
+    }
+
+    return \Civi::service('crypto.jwt')->encode([
+      'iat' => time(),
+      'exp' => time() + 60 * 60 * 24 * $days,
+      'sub' => "user:$pending[userID]",
+      'scope' => 'rememberMe',
+      'hash' => $hashedPass,
+    ]);
+  }
+
+  protected function getPassHash(int $userID): string {
+    // Create a short hash of the hashed password so that when the
+    // user updates their password (even to the same password) then
+    // it invalidates any previously generated rememberJWT values.
+    $hashedPass = User::get(FALSE)
+      ->addWhere('id', '=', $userID)
+      ->addSelect('hashed_password')
+      ->execute()
+      ->first()['hashed_password'] ?? '';
+    if (!$hashedPass) {
+      // This should not happen. But if it does, it's very fishy and we should
+      // not support bypassing MFA.
+      return '';
+    }
+    return hash('xxh3', CIVICRM_SITE_KEY . $hashedPass);
```

Comment:
This feels a bit worrisome - `xxh3` is not a cryptographic hash algorithm. I'm a bit murky on the practical impact and how it applies here given that we're combining a fixed prefix (`CIVICRM_SITE_KEY`) with a kind-of-user-controlled-but-also-salted password hash, I don't think we really need `CIVICRM_SITE_KEY` here.

I think we can get away with a bcrypt of the password hash, or maybe some other cryptographic hash algo. I don't think we should care that much about putting a slightly larger token into `localStorage`, plus that JWT isn't sent to the server all that often.

(An alternative that might have some unrelated benefits could be to keep track of a `password_changed_at` date and compare that to the issuance date of the token, but that's a bit scope-creepy.)

Reply:
So the main thing I'm worried about is the site key and an attacker (who is also a user) possibly being able to recover that.
The hash includes:
- The site key
- followed by a value that is at least partially under the attacker's control (the bcrypt, sans salt)
- ... and which follows a known format (i.e. `$S$...`, fixed length), which might help the attack.

Given that `xxh3` probably doesn't offer any pre-image resistance (i.e. it's easy to find other values resolving to the same hash), and that input *kind of* follows a well-known format, I find it hard to determine if this is a safe construct.

I'm not sure if recovery of the password hash is a real concern here, probably as you described above the threat model for that is different.